lib/oban/queue/executor.ex

```diff
@@ -128,32 +128,46 @@ defmodule Oban.Queue.Executor do
 
   @spec report_finished(t()) :: t()
   def report_finished(%__MODULE__{} = exec) do
-    case exec.state do
-      :success ->
-        Engine.complete_job(exec.conf, exec.job)
+    exec
+    |> ack_event()
+    |> emit_event()
+  end
 
-        execute_stop(exec)
+  @spec ack_event(t()) :: t()
+  def ack_event(%__MODULE__{state: :success} = exec) do
+    Engine.complete_job(exec.conf, exec.job)
+    exec
+  end
 
-      :failure ->
-        job = job_with_unsaved_error(exec)
+  def ack_event(%__MODULE__{state: :failure} = exec) do
+    job = job_with_unsaved_error(exec)
 
-        Engine.error_job(exec.conf, job, backoff(exec.worker, job))
+    Engine.error_job(exec.conf, job, backoff(exec.worker, job))
 
-        execute_exception(%{exec | job: job})
+    %{exec | job: job}
+  end
 
-      :snoozed ->
-        Engine.snooze_job(exec.conf, exec.job, exec.snooze)
+  def ack_event(%__MODULE__{state: :snoozed} = exec) do
+    Engine.snooze_job(exec.conf, exec.job, exec.snooze)
+    exec
+  end
 
-        execute_stop(exec)
+  def ack_event(%__MODULE__{state: :discard} = exec) do
+    job = job_with_unsaved_error(exec)
 
-      :discard ->
-        job = job_with_unsaved_error(exec)
+    Engine.discard_job(exec.conf, job)
 
-        Engine.discard_job(exec.conf, job)
+    %{exec | job: job}
+  end
 
-        execute_stop(%{exec | job: job})
-    end
+  @spec emit_event(t()) :: t()
+  def emit_event(%__MODULE__{state: :failure} = exec) do
+    execute_exception(exec)
```

Comment:
Let's pull the body of `execute_exception` inline. There's no need to put it in a separate private function now.